marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs

```diff
@@ -724,35 +751,62 @@ isAddressInTarget (Just targetAddresses) addr =
-  fromRight Data.Map.empty (getUtxos' txOuts)
+getTxOutFromTxBodyContent :: C.TxBodyContent build era -> [C.TxOut C.CtxTx era]
+getTxOutFromTxBodyContent C.TxBodyContent {..} = case txScriptValidityToScriptValidity  txScriptValidity of
+  C.ScriptValid   -> txOuts -- When transaction is valid, only transaction fee is collected
+  C.ScriptInvalid -> collateral txReturnCollateral -- failed Tx, we collect from collateral and return excess collateral
+  where
+    collateral C.TxReturnCollateralNone       = []
+    collateral (C.TxReturnCollateral _ txout) = [txout]
+
+getUtxosFromTxBody
+  :: (C.IsCardanoEra era)
+  => Maybe TargetAddresses
+  -> C.TxBody era
+  -> Map C.TxIn Utxo
+getUtxosFromTxBody maybeTargetAddresses txBody@(C.TxBody txBodyContent@C.TxBodyContent{} )=
+  fromRight Data.Map.empty (getUtxos' $ getTxOutFromTxBodyContent txBodyContent)
   where
     getUtxos' :: C.IsCardanoEra era => [C.TxOut C.CtxTx era] -> Either C.EraCastError (Map C.TxIn Utxo)
     getUtxos'
       = fmap (Data.Map.fromList . concatMap Data.Map.toList . imap txoutToUtxo)
       . traverse (C.eraCast CurrentEra)
 
+    txid = C.getTxId txBody
     txoutToUtxo :: Int -> TxOut -> Map C.TxIn Utxo
-    txoutToUtxo ix (C.TxOut addr value' datum' refScript) =
-      if isAddressInTarget maybeTargetAddresses addrAny then
-        Data.Map.singleton txin Utxo
-        { _txId = txid
-        , _txIx = txix
-        , _address = addrAny
-        , _value = C.txOutValueToValue value'
-        , _datum = __datum
-        , _datumHash = __datumHash
-        , _inlineScript = __inlineScript
-        , _inlineScriptHash = __inlineScriptHash
-        }
-      else
-        Data.Map.empty
-      where
-        addrAny = toAddr addr
-        (__datum, __datumHash) = getScriptDataAndHash datum'
-        (__inlineScript, __inlineScriptHash) = getRefScriptAndHash refScript
-        txin@(C.TxIn txid txix) = C.TxIn (C.getTxId txBody)(C.TxIx $ fromIntegral ix)
+    txoutToUtxo ix txout =
+      let
+        txin = C.TxIn txid (C.TxIx (fromIntegral ix))
+      in
+        case getUtxoFromTxOut maybeTargetAddresses txin txout of
+          Nothing   -> Data.Map.empty
+          Just utxo -> Data.Map.singleton txin utxo
+
+getUtxoFromTxOut
+  :: Maybe TargetAddresses -- ^ Target addresses to filter for
+  -> C.TxIn -- ^ unique id and position of this transaction
+  -> C.TxOut C.CtxTx era -- ^ Cardano TxOut
+  -> Maybe Utxo -- ^ Utxo
+getUtxoFromTxOut maybeTargetAddresses (C.TxIn txid txix) (C.TxOut addr value' datum' refScript) =
+  if isAddressInTarget maybeTargetAddresses addrAny
+  then Just $ Utxo
+    { _txId = txid
+    , _txIx = txix
+    , _address = addrAny
+    , _value = C.txOutValueToValue value'
+    , _datum = __datum
+    , _datumHash = __datumHash
+    , _inlineScript = __inlineScript
+    , _inlineScriptHash = __inlineScriptHash
+    }
+  else Nothing
+  where
+    addrAny = toAddr addr
+    (__datum, __datumHash) = getScriptDataAndHash datum'
```

Comment:
Maybe not add these underscores?

The idea behind having an underscore in the beginning of a data type field is that it's a function that is not meant to be used directly, but via creating a lens. But in the above we now *do* use all the underscored fields (that we planned access via lens), and then we add double underscores.. Perhaps add a `'` or `_` to signify "temporaryness" of these variables instead?